blog-posts/event-sourcing-serveless/event-sourcing-serveless.md

```diff
@@ -0,0 +1,166 @@
+
+As you can see, the missing part of the diagram is the synchronization between the write model and the read model. That's the block we will focus on.
+
+## First and foremost, in event sourcing, the order of the events is essential
+
+All credit and debit events for every user are stored in the event ledger. The access pattern I chose is the `userId` as the **partition key** and the number of the event also called the `version` as the **sort key**.
+
+![Logo kumo](./assets/event-store-schema.png 'Logo Kumo')
+
+> :question: Why using a `version` field when we can simply take the `timestamp` to keep track of the order of the events?
+
+Two commands coming in **concurrently** will write data based on the **same version of the aggregate** because they fetched it simultaneously. Let's say a user has 50$ on his bank account and two purchases come in at the same time, one of 40$ and one of 25$. Both event would **both succeed the business validation tests**. The composite key `userId`/`timestamp` of both event would be unique because timestamps would be really close but not exactly the same. Hence, both events would be written in the event ledger and you would be **overdrawn** which is **forbidden** in our bank. On the other hand, if the **sort key** is the `version` of the event, **one out of both would be invalid**. Exact simultaneity doesn't exist in computer science, so the first purchase of 40$ would be written and then shortly after the second purchase would be rejected because the event to be written is based on the same version of the last event.
+
+## How is the synchronization between read and write models ensured ?
+
+> The pain point in CQRS is to have **reliable reading data** compared to the source of truth. Projections need to be updated according to events written in the event ledger.
+
+### A fanout listening to Dynamodb Streams emits the events
+
+The objective here is to **react in real time to events** written inside the event ledger in order to compute the projections. For that a lambda as a **fanout** can be used by plugging it to the **DynamoDB streams** of the **event ledger**. `INSERT` actions have to be filtered with the filter patterns so that only theses actions trigger the fanout. DynamoDB batches its streams of events in arrays, so for each of them an event is published in Amazon EventBridge. These EventBridge events are those triggering the lambda that project on the read models.
+
+> :question: You may wonder why not **directly** plug the **projector** to the streams
+
+For 2 reasons:
+
+- Stream events that trigger a failing lambda will retry indefinitely until the lambda succeeds. That makes it **impossible** to have a **custom retry policy** for events triggering a projector.
+- You can only plug **2 lambda maximum** to the streams, and generally application contain more than 2 sets of data to display therefore more than 2 projectors.
```

Comment:
Peut-être ajoute: "The correct tool to plug multiple listeners to messages is an Event Bus"